runtime/vm/ClassInitialization.cpp

```diff
@@ -373,18 +400,80 @@ classInitStateMachine(J9VMThread *currentThread, J9Class *clazz, J9ClassInitStat
 
+					calculateFlattenedFieldAddresses(currentThread, clazz);
+
 					for (UDATA i = 0; i < numberOfFlattenedFields; i++) {
 						J9FlattenedClassCacheEntry *entry = J9_VM_FCC_ENTRY_FROM_CLASS(clazz, i);
-						Trc_VM_classInitStateMachine_verifyFlattenableField(currentThread, entry->clazz);
+						J9Class *entryClazz = NULL;
+						bool isStatic = J9_VM_FCC_ENTRY_IS_STATIC_FIELD(entry);
+
+						if (isStatic) {
+							J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(entry->field);
+							U_8 *signatureChars = J9UTF8_DATA(signature);
+
+							PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, initializationLock);
+							J9Class *valueClass = internalFindClassUTF8(currentThread, signatureChars + 1, J9UTF8_LENGTH(signature) - 2, clazz->classLoader, J9_FINDCLASS_FLAG_THROW_ON_FAIL);
+							initializationLock = POP_OBJECT_IN_SPECIAL_FRAME(currentThread);
+							clazz = VM_VMHelpers::currentClass(clazz);
+							if (NULL == valueClass) {
+								initializationLock = setInitStatus(currentThread, clazz, J9ClassInitUnverified, initializationLock);
+								clazz = VM_VMHelpers::currentClass(clazz);
+								goto done;
+							}
+
+							/* need to check if it has access */
+							J9ROMClass *valueROMClass = valueClass->romClass;
+							bool classIsPublic = J9_ARE_ALL_BITS_SET(valueROMClass->modifiers, J9AccPublic);
+
+							if ((!classIsPublic && (clazz->packageID != valueClass->packageID))
+								|| (classIsPublic && (J9_VISIBILITY_ALLOWED != checkModuleAccess(currentThread, vm, clazz->romClass, clazz->module, valueROMClass, valueClass->module, valueClass->packageID, 0)))
+							) {
+								J9UTF8 *romClassName = J9ROMCLASS_CLASSNAME(valueROMClass);
+								PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, initializationLock);
+								setCurrentExceptionNLSWithArgs(currentThread, J9NLS_VM_CLASS_LOADING_ERROR_INVISIBLE_CLASS_OR_INTERFACE, J9VMCONSTANTPOOL_JAVALANGILLEGALACCESSERROR, J9UTF8_LENGTH(romClassName), J9UTF8_DATA(romClassName));
+								initializationLock = POP_OBJECT_IN_SPECIAL_FRAME(currentThread);
+								clazz = VM_VMHelpers::currentClass(clazz);
+								initializationLock = setInitStatus(currentThread, clazz, J9ClassInitUnverified, initializationLock);
+								clazz = VM_VMHelpers::currentClass(clazz);
+								goto done;
+							}
+
+							entry->clazz = (J9Class *)((UDATA)valueClass | (UDATA)entry->clazz);
+						}
+						entryClazz = J9_VM_FCC_CLASS_FROM_ENTRY(entry);
+
+						if (isStatic) {
+							PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, initializationLock);
+							omrthread_monitor_enter(vm->valueTypeVerificationMutex);
+							BOOLEAN cycleDetected = checkForCyclicalVerification(currentThread, entryClazz->classLoader, entryClazz);
+							omrthread_monitor_enter(vm->valueTypeVerificationMutex);
+							initializationLock = POP_OBJECT_IN_SPECIAL_FRAME(currentThread);
+							clazz = VM_VMHelpers::currentClass(clazz);
+							if (cycleDetected) {
+								continue;
```

Comment:
Why is this not an error?

Reply:
Verification cycles are legal we are verifying static fields. We need to continue to make progress while avoiding the stack overflow

Reply:
ie

```
inline class A { static B b;}
inline class B {static A a;}
```

Reply:
Right, I remember now.

Reply:
Please update the commit/PR comment with some of this detail.

Reply:
And please add a simple comment  in the code to that effect.